r/src/array_to_vector.cpp

```diff
@@ -69,13 +69,25 @@ class Converter {
     // special case when there is only one array
     if (chunked_array_->num_chunks() == 1) {
       const auto& array = chunked_array_->chunk(0);
-      if (arrow::r::GetBoolOption("arrow.use_altrep", true) && array->length() > 0 &&
-          array->null_count() == 0) {
+      // using altrep if
+      // - the arrow.use_altrep is set to TRUE or unset
+      // - the array has at least one element
+      // - either it has no nulls or the data is mutable
+      if (arrow::r::GetBoolOption("arrow.use_altrep", true) && array->length() > 0) {
         switch (array->type()->id()) {
           case arrow::Type::DOUBLE:
-            return arrow::r::MakeDoubleArrayNoNull(array);
+            if (array->null_count() == 0 || array->data()->buffers[1]->is_mutable()) {
```

Comment:
It's not safe to do this if anyone else might mutate this buffer. We don't have a completely safe way to check uniqueness, either. As a stop gap:
```suggestion
            if (array->null_count() == 0 || array->data().use_count() == 1 &&
                array->data()->buffers[1].use_count() == 1 && array->data()->buffers[1]->is_mutable()) {
```

Reply:
@pitrou

Reply:
One consequence is that we can't use altrep when taking slices of an Array as the buffer is at least held by two Arrays then. 

``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

is_altrep <- arrow:::is_altrep

v_int <- Array$create(c(1L, NA, 3L))

is_altrep(v_int$as_vector())
#> [1] TRUE

is_altrep(as.vector(v_int$Slice(1)))
#> [1] FALSE
```

<sup>Created on 2021-07-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

I commented out the relevant tests in `test-altrep.R`. Does this matter given that the data there is irrelevant because it's identified as null ?